src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb

```diff
@@ -306,13 +314,67 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                         End If
                     Loop
 
+                    If (compareKind And TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) = 0 Then
+                        Return HasSameTypeArgumentCustomModifiers(DirectCast(t1, NamedTypeSymbol), DirectCast(t2, NamedTypeSymbol))
+                    End If
+
                     Return True
                 End If
             End If
 
             Return (t1 = t2)
         End Function
 
+        Private Function HasSameTypeArgumentCustomModifiers(type1 As NamedTypeSymbol, type2 As NamedTypeSymbol) As Boolean
+            If Not type1.HasTypeArgumentsCustomModifiers() AndAlso Not type2.HasTypeArgumentsCustomModifiers() Then
+                Return True
+            End If
+
+            Dim mod1 = type1.TypeArgumentsCustomModifiers
+            Dim mod2 = type2.TypeArgumentsCustomModifiers
+
+            If mod1.Length <> mod2.Length Then
+                Return False
+            End If
+
+            If Not mod1.SequenceEqual(mod2, AddressOf AreSameCustomModifiers) Then
+                Return False
+            End If
+
+            Return True
+        End Function
+
+        <Extension()>
+        Friend Function AreSameCustomModifiers([mod] As ImmutableArray(Of CustomModifier), otherMod As ImmutableArray(Of CustomModifier)) As Boolean
+            Dim count As Integer = [mod].Length
+
+            If (count <> otherMod.Length) Then
+                Return False
+            End If
+
+            Return [mod].SequenceEqual(otherMod)
+        End Function
+
+        Private Function HasSameTupleNames(t1 As TypeSymbol, t2 As TypeSymbol) As Boolean
+            If Not t1.IsTupleType AndAlso
+                Not t2.IsTupleType Then
+                Return True
+            End If
+
+            Dim t1Names = t1.TupleElementNames
+            Dim t2Names = t2.TupleElementNames
+
+            If t1Names.IsDefault AndAlso t2Names.IsDefault Then
+                Return True
+            End If
+
+            If t1Names.IsDefault OrElse t2Names.IsDefault OrElse t1Names.Length <> t2Names.Length Then
```

Comment:
Why would we call this method on tuples of different cardinality. It feels like equality of names is not defined for tuples of different cardinality. So, returning ```False``` is likely to mislead the caller and mask a bug elsewhere.  

---
In reply to: [82481901](https://github.com/dotnet/roslyn/pull/14229#discussion_r82481901) [](ancestors = 82481901,82468082)